src/ripple/basics/XRPAmount.h

```diff
@@ -0,0 +1,342 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_XRPAMOUNT_H_INCLUDED
+#define RIPPLE_BASICS_XRPAMOUNT_H_INCLUDED
+
+#include <ripple/basics/contract.h>
+#include <ripple/basics/safe_cast.h>
+#include <ripple/beast/cxx17/type_traits.h>
+#include <ripple/beast/utility/Zero.h>
+#include <ripple/json/json_value.h>
+
+#include <boost/operators.hpp>
+#include <boost/optional.hpp>
+#include <boost/multiprecision/cpp_int.hpp>
+
+#include <cstdint>
+#include <string>
+
+namespace ripple {
+
+namespace feeunit {
+
+/** "drops" are the smallest divisible amount of XRP. This is what most
+    of the code uses. */
+struct dropTag;
+
+} // feeunit
+
+class XRPAmount
+    : private boost::totally_ordered <XRPAmount>
+    , private boost::additive <XRPAmount>
+    , private boost::equality_comparable <XRPAmount, std::int64_t>
+    , private boost::additive <XRPAmount, std::int64_t>
+    , private boost::dividable <XRPAmount, std::int64_t>
+    , private boost::modable <XRPAmount, std::int64_t>
```

Comment:
I have some concerns about implementing multiplication and _significant_ concerns about implementing division in the general purpose class that represents XRP.